SmartDeviceLink/SDLHapticManager.m

```diff
@@ -0,0 +1,161 @@
+
+ @param currentView is the view hierarchy to be processed
+ */
+- (void)sdl_parseViewHierarchy:(UIView *)currentView {
+    if (currentView == nil) {
+        SDLLogW(@"Error: Cannot parse nil view");
+        return;
+    }
+    
+    NSArray *focusableSubviews = [currentView.subviews filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(UIView *  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
+        return evaluatedObject.canBecomeFocused;
+    }]];
+    
+    //if current view is focusable and it doesn't have any focusable sub views then add the cuurent view and return
+    if (currentView.canBecomeFocused && focusableSubviews.count == 0) {
+        [self.focusableViews addObject:currentView];
+        return;
+    }
+    // if current view has focusable sub views parse them recursively
+    else if (currentView.subviews.count > 0) {
+        NSArray<UIView *> *subviews = currentView.subviews;
+        
+        for (UIView *childView in subviews) {
+            [self sdl_parseViewHierarchy:childView];
+        }
+    }
+    //else just return
+    else {
+        return;
+    }
+}
+
+/**
+ Iterates through the focusable views, extracts rectangular parameters, creates Haptic RPC request and sends it
+ */
+- (void)sdl_sendHapticRPC {
+    NSMutableArray<SDLHapticRect *> *hapticRects = [[NSMutableArray alloc] init];
+    
+    for (UIView *view in self.focusableViews) {
+        CGPoint originOnScreen = [view.superview convertPoint:view.frame.origin toView:nil];
+        CGRect convertedRect = {originOnScreen, view.bounds.size};
+        SDLRectangle* rect = [[SDLRectangle alloc] initWithCGRect:(convertedRect)];
+        // using the view index as the id field in SendHapticData request (should be guaranteed unique)
+        NSUInteger rectId = [self.focusableViews indexOfObject:view];
+        SDLHapticRect *hapticRect = [[SDLHapticRect alloc] initWithId:(UInt32)rectId rect:rect];
+        [hapticRects addObject:hapticRect];
+    }
+    
+    SDLSendHapticData* hapticRPC = [[SDLSendHapticData alloc] initWithHapticRectData:hapticRects];
+    [self.connectionManager sendManagerRequest:hapticRPC withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+    }];
+}
+
+#pragma mark SDLHapticHitTester functions
+- (nullable UIView *)viewForSDLTouch:(SDLTouch *)touch {
+    UIView *selectedView = nil;
+    
+    for (UIView *view in self.focusableViews) {
+        //Convert the absolute location to local location and check if that falls within view boundary
+        CGPoint localPoint = [view convertPoint:touch.location fromView:self.projectionWindow];
+        if ([view pointInside:localPoint withEvent:nil]) {
+            if (selectedView != nil) {
+                selectedView = nil;
+                break;
+                //the point has been indentified in two views. We cannot identify which with confidence.
```

Comment:
Returning neither view is what we want here? Is that how `hitTest` works? Is this even possible considering the code in `sdl_parseViewHierarchy` will check subviews? There should be a unit test for this as well.

Reply:
Correct, for the case where two views are overlapping, the hit test will return `nil`.  This would cover the edge case of overlapping views which could be possible for views on the same hierarchy level.  We added a unit test to cover this as well.